To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Unsloth

Load up `Qwen 2.5 1.5B Instruct`, and set parameters

In [ ]:
# Login 
from huggingface_hub import login
login()

In [1]:
from unsloth import FastLanguageModel, is_bfloat16_supported
import wandb
import os 

os.environ["VLLM_USE_V1"] = "0" #https://github.com/unslothai/unsloth/issues/1972

# set the wandb project where this run will be logged
os.environ["WANDB_PROJECT"]="qwen-1.5b-instruct-grpo"

# save your trained model checkpoint to wandb
os.environ["WANDB_LOG_MODEL"]="false"

# turn off watch to log faster
os.environ["WANDB_WATCH"]="false"


max_seq_length = 2048 # Can increase for longer reasoning traces
lora_rank = 64 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/Qwen2.5-1.5B-Instruct",
    max_seq_length = max_seq_length,
    load_in_4bit = True, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.5, # Reduce if out of memory
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ], # Remove QKVO if out of memory
    lora_alpha = lora_rank,
    use_gradient_checkpointing = "unsloth", # Enable long context finetuning
    random_state = 3407,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 05-09 19:40:02 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 05-09 19:40:02 [__init__.py:239] Automatically detected platform cuda.
==((====))==  Unsloth 2025.4.7: Fast Qwen2 patching. Transformers: 4.51.3. vLLM: 0.8.5.post1.
   \\   /|    NVIDIA A100 80GB PCIe. Num GPUs = 1. Max memory: 79.138 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 8.0. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = TRUE. FA [Xformers = 0.0.29.post2. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: vLLM loading unsloth/qwen2.5-1.5b-instruct-unsloth-bnb-4bit with actual GPU utilization = 49.71%
Unsloth: Your GPU has CUDA compute capability 8.0 with VRAM = 79.14 GB.
Unsloth: Using conservativeness

Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


INFO 05-09 19:40:18 [punica_selector.py:18] Using PunicaWrapperGPU.
INFO 05-09 19:40:19 [model_runner.py:1140] Model loading took 1.5709 GiB and 1.741578 seconds
INFO 05-09 19:40:20 [worker.py:287] Memory profiling takes 1.26 seconds
INFO 05-09 19:40:20 [worker.py:287] the current vLLM instance can use total_gpu_memory (79.14GiB) x gpu_memory_utilization (0.50) = 39.34GiB
INFO 05-09 19:40:20 [worker.py:287] model weights take 1.57GiB; non_torch_memory takes 0.09GiB; PyTorch activation peak memory takes 1.75GiB; the rest of the memory reserved for KV Cache is 35.93GiB.
INFO 05-09 19:40:20 [executor_base.py:112] # cuda blocks: 84101, # CPU blocks: 14043
INFO 05-09 19:40:20 [executor_base.py:117] Maximum concurrency for 2048 tokens per request: 657.04x
INFO 05-09 19:40:26 [model_runner.py:1450] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.

Capturing CUDA graph shapes:   0%|          | 0/43 [00:00<?, ?it/s]

INFO 05-09 19:41:00 [model_runner.py:1592] Graph capturing finished in 34 secs, took 0.63 GiB
INFO 05-09 19:41:00 [llm_engine.py:437] init engine (profile, create kv cache, warmup model) took 41.24 seconds


Unsloth 2025.4.7 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


### Data Prep
<a name="Data"></a>

We directly leverage [@willccbb](https://gist.github.com/willccbb/4676755236bb08cab5f4e54a0475d6fb) for data prep and all reward functions. You are free to create your own!

In [3]:
import re
from datasets import load_dataset, Dataset

# Load and prep dataset
SYSTEM_PROMPT = """ You are given a math word problem. You need to solve the problem and provide the answer in the following format: 
<reasoning>
...(explain your reasoning here)...
</reasoning>
<answer>
...(put your answer here)...
</answer>

The answer block should contain only the final answer to the question (i.e., a single number).
"""

XML_COT_FORMAT = """\
<reasoning>
{reasoning}
</reasoning>
<answer>
{answer}
</answer>
"""

def extract_xml_answer(text: str) -> str:
    answer = text.split("<answer>")[-1]
    answer = answer.split("</answer>")[0]
    return answer.strip()

def extract_hash_answer(text: str) -> str | None:
    if "####" not in text:
        return None
    return text.split("####")[1].strip()

# uncomment middle messages for 1-shot prompting
def get_gsm8k_questions(split = "train") -> Dataset:
    data = load_dataset('openai/gsm8k', 'main')[split] # type: ignore
    data = data.map(lambda x: { # type: ignore
        'prompt': [
            {'role': 'system', 'content': SYSTEM_PROMPT},
            {'role': 'user', 'content': x['question']}
        ],
        'answer': extract_hash_answer(x['answer'])
    }) # type: ignore
    return data # type: ignore

dataset = get_gsm8k_questions()

# Reward functions
def correctness_reward_func(prompts, completions, answer, **kwargs) -> list[float]:
    responses = [completion[0]['content'] for completion in completions]
    q = prompts[0][-1]['content']
    extracted_responses = [extract_xml_answer(r) for r in responses]
    print('-'*20, f"Question:\n{q}", f"\nAnswer:\n{answer[0]}", f"\nResponse:\n{responses[0]}", f"\nExtracted:\n{extracted_responses[0]}")
    return [2.0 if r == a else 0.0 for r, a in zip(extracted_responses, answer)]

def int_reward_func(completions, **kwargs) -> list[float]:
    responses = [completion[0]['content'] for completion in completions]
    extracted_responses = [extract_xml_answer(r) for r in responses]
    return [0.5 if r.isdigit() else 0.0 for r in extracted_responses]

def strict_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"^<reasoning>\n.*?\n</reasoning>\n<answer>\n.*?\n</answer>$"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def soft_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"<reasoning>.*?</reasoning>\s*<answer>.*?</answer>"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def count_xml(text) -> float:
    count = 0.0
    if text.count("<reasoning>\n") == 1:
        count += 0.125
    if text.count("\n</reasoning>\n") == 1:
        count += 0.125
    if text.count("\n<answer>\n") == 1:
        count += 0.125
        count -= len(text.split("\n</answer>\n")[-1])*0.001
    if text.count("\n</answer>") == 1:
        count += 0.125
        count -= (len(text.split("\n</answer>")[-1]) - 1)*0.001
    return count

def xmlcount_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    return [count_xml(c) for c in contents]

In [3]:
dataset

Dataset({
    features: ['question', 'answer', 'prompt'],
    num_rows: 7473
})

<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [6]:
from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    use_vllm = True, # use vLLM for fast inference!
    learning_rate = 5e-6,
    adam_beta1 = 0.9,
    adam_beta2 = 0.99,
    weight_decay = 0.1,
    warmup_ratio = 0.1,
    lr_scheduler_type = "cosine",
    optim = "adamw_8bit",
    logging_steps = 5,
    bf16 = is_bfloat16_supported(),
    fp16 = not is_bfloat16_supported(),
    per_device_train_batch_size = 32,
    gradient_accumulation_steps = 4, # Increase to 4 for smoother training
    num_generations = 8, # Decrease if out of memory
    # max_prompt_length = 256,
    max_completion_length = 512,
    num_train_epochs = 1, # Set to 1 for a full training run
    # max_steps = 1000,
    save_steps = 100,
    max_grad_norm = 0.1,
    report_to = "wandb", # Can use Weights & Biases
    output_dir = "outputs",
)

And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [7]:
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        xmlcount_reward_func,
        soft_format_reward_func,
        strict_format_reward_func,
        int_reward_func,
        correctness_reward_func,
    ],
    args = training_args,
    train_dataset = dataset,
)


In [8]:
trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 7,473 | Num Epochs = 1 | Total steps = 467
O^O/ \_/ \    Batch size per device = 32 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (32 x 4 x 1) = 128
 "-____-"     Trainable parameters = 73,859,072/5,000,000,000 (1.48% trained)
wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
wandb: Currently logged in as: lukechen526 to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


-------------------- Question:
A concert ticket costs $40. Mr. Benson bought 12 tickets and received a 5% discount for every ticket bought that exceeds 10. How much did Mr. Benson pay in all? 
Answer:
476 
Response:
Mr. Benson bought 12 tickets, so he bought 12 - 10 = 2 tickets that exceed 10. For these 2 tickets, he gets a 5% discount, so he saves 2 * 0.05 = $1 on these tickets.

So 12 tickets cost him 12 * $40 = $480.

And the 2 tickets that are over have a savings of $1 each, so he saves a total of 2 * $1 = $2.

So in total, Mr. Benson paid $480 - $2 = $478 for the tickets. 
Extracted:
Mr. Benson bought 12 tickets, so he bought 12 - 10 = 2 tickets that exceed 10. For these 2 tickets, he gets a 5% discount, so he saves 2 * 0.05 = $1 on these tickets.

So 12 tickets cost him 12 * $40 = $480.

And the 2 tickets that are over have a savings of $1 each, so he saves a total of 2 * $1 = $2.

So in total, Mr. Benson paid $480 - $2 = $478 for the tickets.
-------------------- Question:
Over 

-------------------- Question:
Zach baked cookies for a bake sale. On Monday he baked 32 cookies. On Tuesday he ran out of sugar and could only bake half of the number of cookies he baked the day before. On Wednesday he baked three times the number of cookies he did on Tuesday, but his brother ate 4 of those cookies. How many cookies did he have at the end of the three days? 
Answer:
92 
Response:
Reasoning:
1. On Monday, Zach baked 32 cookies.
2. On Tuesday, Zach baked half of the number of cookies he baked on Monday, so that's 32 / 2 = 16 cookies. But he ran out of sugar, so he didn't bake any more cookies.
3. On Wednesday, Zach baked three times the number of cookies he did on Tuesday, so that's 3 * 16 = 48 cookies. However, his brother ate 4 of those cookies, so he has 48 - 4 = 44 cookies left.
4. To find out how many cookies he had at the end of the three days, we need to add the cookies he had from each day together. So, we have: 32 (Monday) + 0 (Tuesday) + 44 (Wednesday) = 32 + 

TrainOutput(global_step=467, training_loss=0.004250276013365592, metrics={'train_runtime': 9980.6525, 'train_samples_per_second': 0.749, 'train_steps_per_second': 0.047, 'total_flos': 0.0, 'train_loss': 0.004250276013365592})

In [9]:
# [optional] finish the wandb run, necessary in notebooks
wandb.finish()

<a name="Inference"></a>
### Inference
Now let's try the model we just trained! First, let's first try the model without any GRPO trained:

In [24]:
dataset_test = get_gsm8k_questions("test").shuffle(seed=42).select(range(30))

In [25]:
tokenized = tokenizer.apply_chat_template(dataset_test['prompt'], tokenize=False, add_generation_prompt=True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
generated = model.fast_generate(
    tokenized,
    sampling_params = sampling_params,
    lora_request = None,
)

completions = [g.outputs[0].text for g in generated]

score = sum([1.0 if extract_xml_answer(c) == answer else 0.0 for c, answer in zip(completions, dataset_test['answer'])])
print(f"Accuracy: {score / len(dataset_test) * 100:.2f}%")

# Print the generated completions and the answers
for i, (completion, answer) in enumerate(zip(completions, dataset_test['answer'])):
    print(f"Completion {i+1}: \n {completion}")
    print(f"Correct Answer: {answer}")
    print("-" * 50)

Processed prompts:   0%|          | 0/30 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

Accuracy: 16.67%
Completion 1: 
 The ratio of Darrell's age to Allen's age is 7:11. This means that for every 7 parts of Darrell's age, there are 11 parts of Allen's age.

Let's denote the total parts as 18 (since 7 + 11 = 18).

Now, let's calculate the value of each part:

\[ \text{Value of each part} = \frac{162}{18} = 9 \]

Now, we can find Allen's current age by multiplying the value of each part by the number of parts for Allen:

\[ \text{Allen's current age} = 9 \times 11 = 99 \]

To calculate Allen's age 10 years from now, we add 10 to his current age:

\[ \text{Allen's age 10 years from now} = 99 + 10 = 109 \]

Therefore, the answer is 109.
Correct Answer: 109
--------------------------------------------------
Completion 2: 
 <reasoning>
First, calculate the number of large buttons Lorraine receives from trading large stickers for small buttons. Lorraine trades 50% of her large stickers, which is \(0.5 \times 40 = 20\) large stickers. Since each large sticker is worth a large b

And now with the LoRA we just trained with GRPO - we first save the LoRA first!

In [12]:
model.save_lora("grpo_saved_lora")

Now we load the LoRA and test:

In [26]:
tokenized = tokenizer.apply_chat_template(dataset_test['prompt'], tokenize=False, add_generation_prompt=True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
generated = model.fast_generate(
    tokenized,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)

completions = [g.outputs[0].text for g in generated]

score = sum([1.0 if extract_xml_answer(c) == answer else 0.0 for c, answer in zip(completions, dataset_test['answer'])])
print(f"Accuracy: {score / len(dataset_test) * 100:.2f}%")
# Print the generated completions and the answers
for i, (completion, answer) in enumerate(zip(completions, dataset_test['answer'])):
    print(f"Completion {i+1}: \n {completion}")
    print(f"Correct Answer: {answer}")
    print("-" * 50)

Processed prompts:   0%|          | 0/30 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

Accuracy: 46.67%
Completion 1: 
 <reasoning>
Let's denote Darrell's age as \( 7x \) and Allen's age as \( 11x \). The total age is given as 162, so we can write the equation \( 7x + 11x = 162 \). Simplifying this, we get \( 18x = 162 \), and solving for \( x \) gives \( x = 9 \). Therefore, Darrell's age is \( 7 \times 9 = 63 \) and Allen's age is \( 11 \times 9 = 99 \). To find Allen's age 10 years from now, we add 10 to his current age: \( 99 + 10 = 109 \).
</reasoning>
<answer>
109
</answer>
Correct Answer: 109
--------------------------------------------------
Completion 2: 
 <reasoning>
Lorraine has 30 small stickers and 40 large stickers. She trades 90% of her small stickers (27) for large buttons, and 50% of her large stickers (20) for large buttons, which means she trades the remaining 50% of her large stickers (20) for small buttons. She started with 30 small stickers and traded 27 for large buttons, leaving her with 3 small stickers. She then trades 20 large buttons for 60 sm

Our reasoning model is much better - it's not always correct, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [27]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if True: model.push_to_hub_merged("qwen2.5-1.5b-gsm8k-grpo-finetune", tokenizer, save_method = "merged_16bit")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("qwen2.5-1.5b-gsm8k-grpo-finetune-lora", tokenizer, save_method = "lora")
if False: model.push_to_hub_merged("qwen2.5-1.5b-gsm8k-grpo-finetune-lora", tokenizer, save_method = "lora")

Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 67.58 out of 118.01 RAM for saving.
Unsloth: Saving model... This might take 5 minutes ...


100%|██████████| 28/28 [00:00<00:00, 125.06it/s]


Unsloth: Saving to organization with address lukechen526/qwen2.5-1.5b-gsm8k-grpo-finetune
Unsloth: Saving tokenizer... Done.
Unsloth: Saving to organization with address lukechen526/qwen2.5-1.5b-gsm8k-grpo-finetune
Unsloth: Uploading all files... Please wait...


  0%|          | 0/2 [00:00<?, ?it/s]

model.safetensors:   0%|          | 0.00/3.09G [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

Done.
Saved merged model to https://huggingface.co/None/qwen2.5-1.5b-gsm8k-grpo-finetune


### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("qwen2.5-1.5b-gsm8k-grpo-finetune-gguf", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "qwen2.5-1.5b-gsm8k-grpo-finetune-gguf", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",]
    )

Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 76.27 out of 118.01 RAM for saving.
Unsloth: Saving model... This might take 5 minutes ...


100%|██████████| 28/28 [00:00<00:00, 197.48it/s]

Unsloth: Saving tokenizer...

 Done.
Done.
==((====))==  Unsloth: Conversion from QLoRA to GGUF information
   \\   /|    [0] Installing llama.cpp might take 3 minutes.
O^O/ \_/ \    [1] Converting HF to GGUF 16bits might take 3 minutes.
\        /    [2] Converting GGUF 16bits to ['f16'] might take 10 minutes each.
 "-____-"     In total, you will have to wait at least 16 minutes.

Unsloth: Installing llama.cpp. This might take 3 minutes...


RuntimeError: Unsloth: The file ('llama.cpp/llama-quantize' or 'llama.cpp/llama-quantize.exe' if you are on Windows WSL) or 'llama.cpp/quantize' does not exist.
But we expect this file to exist! Maybe the llama.cpp developers changed the name or check extension of the llama-quantize file.

## Load from Hugging Face

In [1]:
from unsloth import FastLanguageModel, is_bfloat16_supported
import wandb
import os 

os.environ["VLLM_USE_V1"] = "0" #https://github.com/unslothai/unsloth/issues/1972


max_seq_length = 2048 # Can increase for longer reasoning traces
lora_rank = 64 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "lukechen526/qwen2.5-1.5b-gsm8k-grpo-finetune-lora",
    max_seq_length = max_seq_length,
    fast_inference = True, # Enable vLLM fast inference
)




🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 05-10 03:08:39 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 05-10 03:08:39 [__init__.py:239] Automatically detected platform cuda.
==((====))==  Unsloth 2025.4.7: Fast Qwen2 patching. Transformers: 4.51.3. vLLM: 0.8.5.post1.
   \\   /|    NVIDIA GeForce RTX 3080. Num GPUs = 1. Max memory: 10.0 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 8.6. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = TRUE. FA [Xformers = 0.0.29.post2. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: vLLM loading unsloth/qwen2.5-1.5b-instruct-unsloth-bnb-4bit with actual GPU utilization = 44.15%
Unsloth: Your GPU has CUDA compute capability 8.6 with VRAM = 10.0 GB.
Unsloth: Using conservativeness 

Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


INFO 05-10 03:08:56 [punica_selector.py:18] Using PunicaWrapperGPU.
INFO 05-10 03:08:57 [model_runner.py:1140] Model loading took 1.5709 GiB and 3.660636 seconds
INFO 05-10 03:08:58 [worker.py:287] Memory profiling takes 1.27 seconds
INFO 05-10 03:08:58 [worker.py:287] the current vLLM instance can use total_gpu_memory (10.00GiB) x gpu_memory_utilization (0.44) = 4.41GiB
INFO 05-10 03:08:58 [worker.py:287] model weights take 1.57GiB; non_torch_memory takes 0.05GiB; PyTorch activation peak memory takes 0.88GiB; the rest of the memory reserved for KV Cache is 1.92GiB.
INFO 05-10 03:08:58 [executor_base.py:112] # cuda blocks: 4485, # CPU blocks: 9362
INFO 05-10 03:08:58 [executor_base.py:117] Maximum concurrency for 2048 tokens per request: 35.04x
INFO 05-10 03:08:59 [model_runner.py:1450] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI. If o

Capturing CUDA graph shapes:   0%|          | 0/23 [00:00<?, ?it/s]

INFO 05-10 03:09:16 [model_runner.py:1592] Graph capturing finished in 18 secs, took 0.38 GiB
INFO 05-10 03:09:16 [llm_engine.py:437] init engine (profile, create kv cache, warmup model) took 19.93 seconds


Unsloth 2025.4.7 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


In [6]:
model.save_lora("grpo_saved_lora")

In [31]:
dataset_test = get_gsm8k_questions("test").shuffle(seed=42).select(range(50))

tokenized = tokenizer.apply_chat_template(dataset_test['prompt'], tokenize=False, add_generation_prompt=True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
generated = model.fast_generate(
    tokenized,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
    # lora_request = None,
)

completions = [g.outputs[0].text for g in generated]

score = sum([1.0 if extract_xml_answer(c) == answer else 0.0 for c, answer in zip(completions, dataset_test['answer'])])
print(f"Accuracy: {score / len(dataset_test) * 100:.2f}%")

# Print the generated completions and the answers
for i, (completion, answer) in enumerate(zip(completions, dataset_test['answer'])):
    print(f"Completion {i+1}: \n{completion}")
    print(f"Correct Answer: {answer}")
    print("-" * 50)

Processed prompts:   0%|          | 0/50 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

Accuracy: 58.00%
Completion 1: 
<reasoning>
Let's denote Darrell's age as \(7x\) and Allen's age as \(11x\). Their total age is given by the equation \(7x + 11x = 162\), which simplifies to \(18x = 162\). Solving for \(x\), we get \(x = 9\). Therefore, Darrell's age is \(7x = 63\) and Allen's age is \(11x = 99\). In 10 years, Allen's age will be \(99 + 10 = 109\).
</reasoning>
<answer>
109
</answer>
Correct Answer: 109
--------------------------------------------------
Completion 2: 
<reasoning>
Lorraine starts with 30 small stickers. She trades 90% of them, i.e., 27 small stickers, for large buttons. Since each large button is worth 3 small stickers, she will receive 27/3 = 9 large buttons. She also trades 50% of her large stickers, i.e., 20 large stickers, for large buttons. So, she will receive 20/3 ≈ 6.67 large buttons, but since she can only trade whole large stickers, she will receive 6 large buttons. She also trades the rest of her large stickers, i.e., 20 large stickers, for la

In [30]:
text = tokenizer.apply_chat_template([
    {"role" : "system", "content" : SYSTEM_PROMPT},
    {"role" : "user", "content" : "James decides to run 3 sprints 3 times a week. He runs 60 meters each sprint. How many total meters does he run a week?"},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

print(output)

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

<reasoning>
James runs 3 sprints each time, and each sprint is 60 meters long. So, he runs \(3 \times 60 = 180\) meters each time. Since he runs 3 times a week, the total distance he runs in a week is \(180 \times 3 = 540\) meters.
</reasoning>
<answer>
540
</answer>


Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
